### PR TITLE
feat(#22): Setup markdown editor

### PR DESCRIPTION
**Pull request by ddsuhaimi**

### CODE REVIEW

File: client/src/services/injectElipsis.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    let lastMatchIndex;
+    let match;
+    while ((match = re.exec(str)) != null) {
+        lastMatchIndex = match.index;
+    }
+
+    return html.substring(0, lastMatchIndex) + "..." + html.substring(lastMatchIndex)
+}
+
+export default injectElipsis;
```
Comment (Mayank0255) [inline client/src/services/injectElipsis.js:13]: I think there is a typo in the name `injectElipsis` which should be `injectEllipsis`. Kindly change the file name, the variable name, export name, and wherever imported.

File: client/src/components/RichTextEditor/draft.css [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
...x}.public-DraftStyleDefault-orderedListItem.public-DraftStyleDefault-listRTL:before{position:absolute;right:-36px;text-align:left;width:30px}.public-DraftStyleDefault-orderedListItem:before{content:counter(ol0) ". ";counter-increment:ol0}.public-DraftStyleDefault-orderedListItem.public-DraftStyleDefault-depth1:before{content:counter(ol1,lower-alpha) ". ";counter-increment:ol1}.public-DraftStyleDefault-orderedListItem.public-DraftStyleDefault-depth2:before{content:counter(ol2,lower-roman) ". ";counter-increment:ol2}.public-DraftStyleDefault-orderedListItem.public-DraftStyleDefault-depth3:before{content:counter(ol3) ". ";counter-increment:ol3}.public-DraftStyleDefault-orderedListItem.public-DraftStyleDefault-depth4:before{content:counter(ol4,lower-alpha) ". ";counter-increment:ol4}.public-DraftStyleDefault-depth0.public-DraftStyleDefault-reset{counter-reset:ol0}.public-DraftStyleDefault-depth1.public-DraftStyleDefault-reset{counter-reset:ol1}.public-DraftStyleDefault-depth2.public-DraftStyleDefault-reset{counter-reset:ol2}.public-DraftStyleDefault-depth3.public-DraftStyleDefault-reset{counter-reset:ol3}.public-DraftStyleDefault-depth4.public-DraftStyleDefault-reset{counter-reset:ol4}
```
Comment (Mayank0255) [inline client/src/components/RichTextEditor/draft.css:10]: Its in minified format of `css`, I would recommend formatting it properly and then pasting.

For help - https://www.freeformatter.com/css-beautifier.html

**Review by Mayank0255 — CHANGES_REQUESTED** [review]

**Comment by Mayank0255** [comment]
@ddsuhaimi I have requested some small changes on the PR, other than that great work.

**Comment by ddsuhaimi** [comment]
@Mayank0255 
Updated with the requested changes

**Comment by Mayank0255** [comment]
@ddsuhaimi 

Great job man.

I just checked it out and it works fine for the functionalities given but you have forgotten to add __code__ and __hyperlink__ functionality to it. While adding the code one don't forget to look at the ui of original stackoverflow clone.

And one doubt I had is that what is `htmlSubstring` used for with `injectEllipsis`?

**Comment by Mayank0255** [comment]
@ddsuhaimi One last thing that don't forget to remove the monospace feature as it's not required

**Comment by ddsuhaimi** [comment]
@Mayank0255 
For htmlSubstring and injectEllipsis, I used it for post item in the homepage. Normally, you substring the body to show only first 200 characters right?
Since what we have right now is a string of html, I can't simply do body.substring(200), so htmlSubstring comes to play. What it does is that it will extract the first 200 character without touching the html tag (so we wont lose any formatting). I then created injectEllipsis to inject ellipsis at the end of that substring. If I dont do that, when it happen to cut off in the middle of the word, it looks weird.

For the other 2, the converter library still having an open issues for code block (https://github.com/HubSpot/draft-convert/issues/182) and I'll see how to implement the link since it doesn't come by default in draft.js

**Comment by Mayank0255** [comment]
> For htmlSubstring and injectEllipsis, I used it for post item in the homepage. Normally, you substring the body to show only first 200 characters right?

Yes, good work 👍🏽 

> I'll see how to implement the link since it doesn't come by default in draft.js

Sure, do that please and work on the code block thing and removing the monospace too while you are at it.

**Comment by Mayank0255** [comment]
@ddsuhaimi 

What's the progress on this?

**Comment by ddsuhaimi** [comment]
@ddsuhaimi 

Sorry, it is probably gonna take a while. I've been kinda busy lately. I'll develop this in the weekend.

**Comment by Mayank0255** [comment]
@ddsuhaimi Sure

**Comment by Mayank0255** [comment]
@ddsuhaimi Are you facing any kind of problem somewhere?

**Comment by Mayank0255** [comment]
@ddsuhaimi Are you facing any kind of problem somewhere?

**Comment by ddsuhaimi** [comment]
@Mayank0255 

Hi, sorry it took a while. Turns out, the issue is on 'draft-convert' library itself. It just can't convert code block to its corresponding html (and it is actually an open issue in its repo and still hasn't been resolved). So, there is pretty much no use of having code block in the editor when we can't convert it. This is also the case for link.

**Comment by Mayank0255** [comment]
@ddsuhaimi 

Okay, I see

Here's an answer for draftjs to html conversion by the author himself - [Here](https://stackoverflow.com/questions/35884112/draftjs-how-to-initiate-an-editor-with-content/36007299#36007299)

In which the author has recommended to use the libraries listed over there as it's officially built by them and for a running example you can see [this](https://react-rte.org/demo) and the repository to which is listed [here](https://github.com/sstur/react-rte)

In the demo shown __code block__ is specifically implemented also. The link to the package repo is [here](https://github.com/sstur/draft-js-utils)

I hope this solves the issue and if not then let me know.

I know this can be tiring to find out the solution but till know you have done a marvelous job so it won't hurt to finish it off perfectly too. It might seem a big task of replacing the draft-convert with this but it would take hardly and hour after figuring out the approach and for which I am here.

P.S. - I am thinking of integrating docusaurus to the backend in which along with the API, a blog page would also be there in which news would come out referring the contributors and congratulating them and explaining about the contribution to the project. This is just to tell you that the contribution won't be in vain.

**Comment by ddsuhaimi** [comment]
@Mayank0255 

so is it okay if instead of using pure draft-js with draft-convert, we use react-rte? which i think, under the hood, is draft-js?

**Comment by Mayank0255** [comment]
@ddsuhaimi Sure, go ahead

As long as it satisfies our needs, its fine

**Review by Mayank0255 — COMMENTED** [review]
@ddsuhaimi 

The code looks pretty much fine, so for better review and my convenience could you add two screenshots?
1. of the editor with information written in it while applying most of the essential features, like bold, italics, underline, code block, lists, hyperlinks, etc.
2. of the rendered output of the same in the Post page of the particular post you added

Would that be fine with you to attach?

**Comment by ddsuhaimi** [comment]
@Mayank0255 
Sure thing. 

image
image

**Comment by Mayank0255** [comment]
@ddsuhaimi Good job man!!

Thank you for working patiently with me 😄 

Looking forward to more of the contributions from your side

**Pull request merged**